File: app/src/main/java/com/adsamcik/signalcollector/data/Challenge.java
package com.adsamcik.signalcollector.data;

import android.content.Context;
import android.content.res.Resources;
import android.support.annotation.NonNull;
import android.util.Log;

import com.adsamcik.signalcollector.R;
import com.adsamcik.signalcollector.enums.ChallengeDifficulty;
import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;
import com.vimeo.stag.UseStag;

public class Challenge {
	/**
	 * Type of challenge
	 */
	private ChallengeType type;

	/**
	 * Challenge progress
	 */
	private float progress;

	/**
	 * Description variables used to format description strings
	 */
	private String[] descVars;

	/**
	 * Difficulty of the challenge
	 */
	private @ChallengeDifficulty int difficulty;

	private String title;
	private String description;
	private String difficultyString;

	public Challenge(ChallengeType type, String title, String description, float progress, @ChallengeDifficulty int difficulty) {
		this.type = type;
		this.title = title;
		this.description = description;
		this.descVars = null;
		this.progress = progress;
		this.difficulty = difficulty;
	}

	public Challenge(ChallengeType type, String title, String[] descVars, float progress, @ChallengeDifficulty int difficulty) {
		this.type = type;
		this.title = title;
		this.descVars = descVars;
		this.description = null;
		this.progress = progress;
		this.difficulty = difficulty;
	}

	public ChallengeType getType() {
		return type;
	}

	public String getTitle() {
		return title;
	}

	public String getDescription() {
		return description;
	}

	public String getDifficultyString() {
		return difficultyString;
	}

	public float getProgress() {
		return progress;
	}

	public void setDone(boolean done) {
		this.progress = 1;
	}

	public void generateTexts(@NonNull Context context) {
		Resources resources = context.getResources();
		assert descVars != null;
		switch (type) {
			case LawfulExplorer:
				title = resources.getString(R.string.challenge_lawful_explorer_title);
				description = resources.getString(R.string.challenge_lawful_explorer_description, descVars[0], descVars[1]);
				break;
			case AwfulExplorer:
				title = resources.getString(R.string.challenge_awful_explorer_title);
				description = resources.getString(R.string.challenge_awful_explorer_description, descVars[0]);
				break;
			case ScatteredData:
				title = resources.getString(R.string.challenge_scattered_data_title);
				description = resources.getString(R.string.challenge_scattered_data_description, descVars[0]);
				break;
			case LongJourney:
				title = resources.getString(R.string.challenge_long_journey_title);
				description = resources.getString(R.string.challenge_long_journey_description, descVars[0]);
				break;
			case Crowded:
				title = resources.getString(R.string.challenge_crowded_title);
				description = resources.getString(R.string.challenge_crowded_description, descVars[0]);
				break;
			case PeacefulForest:
				title = resources.getString(R.string.challenge_peaceful_forest_title);
				description = resources.getString(R.string.challenge_peaceful_forest_description, descVars[0]);
				break;
			case CrossCountry:
				title = resources.getString(R.string.challenge_cross_country_title);
				description = resources.getString(R.string.challenge_cross_country_description, descVars[0]);
				break;
			case TripDownMemoryLane:
				title = resources.getString(R.string.challenge_memory_lane_title);
				description = resources.getString(R.string.challenge_memory_lane_description, descVars[0]);
				break;
			case AIsForAlphabet:
				title = resources.getString(R.string.challenge_alphabet_title);
				description = resources.getString(R.string.challenge_alphabet_description, descVars[0], descVars[1]);
				break;
			case StayInRange:
				title = resources.getString(R.string.challenge_cell_range_title);
				description = resources.getString(R.string.challenge_cell_range_description, descVars[0]);
				break;
			default:
				description = resources.getString(R.string.challenge_error_name);
				break;
		}

		switch (difficulty) {
			case ChallengeDifficulty.EASY:
				difficultyString = resources.getString(R.string.challenge_easy);
				break;
			case ChallengeDifficulty.HARD:
				difficultyString = resources.getString(R.string.challenge_hard);
				break;
			case ChallengeDifficulty.MEDIUM:
				difficultyString = resources.getString(R.string.challenge_medium);
				break;
			case ChallengeDifficulty.VERY_EASY:
				difficultyString = resources.getString(R.string.challenge_very_easy);
				break;
			case ChallengeDifficulty.VERY_HARD:
				difficultyString = resources.getString(R.string.challenge_very_hard);
				break;
			default:
				difficultyString = null;
				break;
		}
	}

	public enum ChallengeType {
		LawfulExplorer,
		AwfulExplorer,
		ScatteredData,
		LongJourney,
		Crowded,
		PeacefulForest,
		CrossCountry,
		TripDownMemoryLane,
		AIsForAlphabet,
		StayInRange
	}
}

File: app/src/main/java/com/adsamcik/signalcollector/data/Challenge.kt
package com.adsamcik.signalcollector.data

import android.content.Context
import com.adsamcik.signalcollector.R
import com.adsamcik.signalcollector.enums.ChallengeDifficulties

/**
 * Challenge object that holds information about challenge including it's text localization
 */

class Challenge {
    /**
     * Type of challenge
     */
    var type: ChallengeType? = null
        private set

    /**
     * Challenge progress
     */
    var progress: Float = 0.toFloat()
        private set

    /**
     * Description variables used to format description strings
     */
    private var descVars: Array<String>? = null

    /**
     * Difficulty of the challenge
     */
    @ChallengeDifficulties.ChallengeDifficulty
    private var difficulty: Int = 0

    var title: String? = null
        private set
    var description: String? = null
        private set
    var difficultyString: String? = null
        private set

    constructor()

    constructor(type: ChallengeType, title: String, description: String, progress: Float, @ChallengeDifficulties.ChallengeDifficulty difficulty: Int) {
        this.type = type
        this.title = title
        this.description = description
        this.descVars = null
        this.progress = progress
        this.difficulty = difficulty
    }

    constructor(type: ChallengeType, title: String, descVars: Array<String>, progress: Float, @ChallengeDifficulties.ChallengeDifficulty difficulty: Int) {
        this.type = type
        this.title = title
        this.descVars = descVars
        this.description = null
        this.progress = progress
        this.difficulty = difficulty
    }

    /**
     * Sets progress as done
     * This is useful when progress needs to be updated locally when challenge is completed without redownloading challenges from the server.
     */
    fun setDone() {
        this.progress = 1f
    }

    /**
     * This function loads appropriate strings from the resources for this challenge.
     * Implementing it like this allowed proper localization.
     */
    fun generateTexts(context: Context) {
        val resources = context.resources
        assert(descVars != null)
        when (type) {
            Challenge.ChallengeType.LawfulExplorer -> {
                title = resources.getString(R.string.challenge_lawful_explorer_title)
                description = resources.getString(R.string.challenge_lawful_explorer_description, descVars!![0], descVars!![1])
            }
            Challenge.ChallengeType.AwfulExplorer -> {
                title = resources.getString(R.string.challenge_awful_explorer_title)
                description = resources.getString(R.string.challenge_awful_explorer_description, descVars!![0])
            }
            Challenge.ChallengeType.ScatteredData -> {
                title = resources.getString(R.string.challenge_scattered_data_title)
                description = resources.getString(R.string.challenge_scattered_data_description, descVars!![0])
            }
            Challenge.ChallengeType.LongJourney -> {
                title = resources.getString(R.string.challenge_long_journey_title)
                description = resources.getString(R.string.challenge_long_journey_description, descVars!![0])
            }
            Challenge.ChallengeType.Crowded -> {
                title = resources.getString(R.string.challenge_crowded_title)
                description = resources.getString(R.string.challenge_crowded_description, descVars!![0])
            }
            Challenge.ChallengeType.PeacefulForest -> {
                title = resources.getString(R.string.challenge_peaceful_forest_title)
                description = resources.getString(R.string.challenge_peaceful_forest_description, descVars!![0])
            }
            Challenge.ChallengeType.CrossCountry -> {
                title = resources.getString(R.string.challenge_cross_country_title)
                description = resources.getString(R.string.challenge_cross_country_description, descVars!![0])
            }
            Challenge.ChallengeType.TripDownMemoryLane -> {
                title = resources.getString(R.string.challenge_memory_lane_title)
                description = resources.getString(R.string.challenge_memory_lane_description, descVars!![0])
            }
            Challenge.ChallengeType.AIsForAlphabet -> {
                title = resources.getString(R.string.challenge_alphabet_title)
                description = resources.getString(R.string.challenge_alphabet_description, descVars!![0], descVars!![1])
            }
            Challenge.ChallengeType.StayInRange -> {
                title = resources.getString(R.string.challenge_cell_range_title)
                description = resources.getString(R.string.challenge_cell_range_description, descVars!![0])
            }
            else -> description = resources.getString(R.string.challenge_error_name)
        }

        difficultyString = when (difficulty) {
            ChallengeDifficulties.EASY -> resources.getString(R.string.challenge_easy)
            ChallengeDifficulties.HARD -> resources.getString(R.string.challenge_hard)
            ChallengeDifficulties.MEDIUM -> resources.getString(R.string.challenge_medium)
            ChallengeDifficulties.VERY_EASY -> resources.getString(R.string.challenge_very_easy)
            ChallengeDifficulties.VERY_HARD -> resources.getString(R.string.challenge_very_hard)
            ChallengeDifficulties.UNKNOWN -> null
            else -> null
        }
    }

    /**
     * Challenge types that are currently supported
     */
    enum class ChallengeType {
        LawfulExplorer,
        AwfulExplorer,
        ScatteredData,
        LongJourney,
        Crowded,
        PeacefulForest,
        CrossCountry,
        TripDownMemoryLane,
        AIsForAlphabet,
        StayInRange
    }
}
